Hubverse ms vignette and associated documents

nickreich opened the pull request:
**Review type**: In-depth
**Review timeline**: Not blocking immediate progress, but by Wednesday, June 18 would be nice so we can continue with manuscript submission, or make changes as needed.

This is a version of the vignette that will be submitted as part of the forthcoming hubverse MS. I ran a 
`remotes::install_github("hubverse-org/hubverse")`
before running the vignette, so hopefully I have the most recent versions of all the things. Some specific things for reviewers to look for:

 - can you run this on your machine without issue?
 - is there a better way to load the target data yet? Currently, I have some custom code reading an S3 file directly.

I tried to fix the lintr warnings, but not the other CI warnings.

nickreich committed:
adding draft of hubverse MS vignette.

nickreich committed:
add pdf

nickreich committed:
change color of models in ensemble plots, using new hubVis.

nickreich committed:
updating lintr complaints and adding pdf.

eahowerton commented on vignettes/hubverse_ms_vignette.Rmd:
```diff
@@ -0,0 +1,220 @@
+
+The package hubEvals can be used to evaluate probabilistic predictions against observed data. The hubEvals package creates an interface for hubverse-style model output and target data to the scoringutils package, which is an engine for computing a number of common scores for probabilistic predictions.[@bosse-scoringutils-2024] 
+
+The code below starts by making oracle output data, which is a dataset formatted like hubverse-style model output data but with a point prediction representing the eventual observation, as if observed by an oracle model. The FluSight hub stores target data in a time-series format but not in an oracle output format. The oracle output data format is convenient to use for scoring, as it can be joine ...
```
I think this should be "average weighted interval score"

LucieContamin replied:
Should we add a little bit more, by saying that you are using the `hubEvals::score_model_out()` function that scores model outputs with a single output_type against observed data, and more information is available in the help page of the function and the associated package?

eahowerton reviewed (dismissed):
This looks really nice! It's a concise and easy to use overview of the hubverse tools. I was able to compile the vignette on my machine without issue.

I've left a few very minor comments for consideration. The only comments that should be addressed before merging are related to Figure 2.

I also have one bigger-picture thought. There are some key hubverse concepts throughout the vignette (e.g., "oracle data format") that are used but not defined. I don't think this is a problem per se, and this vignette doesn't seem like the place to define them in detail. However, it does mean that the vignette doesn't stand on its own for an unfamiliar reader. A few ideas of ways to address this without going into too much depth: 
1. A sentence or two at the beginning of the vignette pointing readers to either the website or the paper for an overview of key concepts
2. Link to specific vignettes or pages on the website where relevant concepts are introduced. 

This is only a suggestion - you have a much better idea of whether it would be useful for this vignette to be stand-alone beyond just a supplement to the paper

LucieContamin commented on vignettes/hubverse_ms_vignette.Rmd:
````diff
@@ -0,0 +1,220 @@
+
+
+```{r, echo=FALSE}
+knitr::kable(flusight_data[116:120,
+                           c("model_id", "reference_date", "location",
+                             "output_type", "output_type_id", "value")],
+             digits = 2,
+             caption = "Five rows of the `flusight_data` object collected above.
+             The \"target\", \"horizon\", and \"target_end_date\" columns have
+             been excluded for readability. The \"output_type_id\" and \"value\"
+             columns show the different predicted quantiles for the number of
+             hospital admissions due to influenza made by the CMU-TimeSeries
+             model for location 48 (Texas). The predictions were made during the
+             week of December 16, 2023 (reference date) and are making a
+             prediction for the horizon of 1 week ahead, for the week ending
+             December 23.")
````
```suggestion
knitr::kable(flusight_data[116:120,
                           c("model_id", "reference_date", "horizon",
                             "output_type", "output_type_id", "value")],
             digits = 2,
             caption = "Five rows of the `flusight_data` object collected above.
             The \"target\", \"location\", and \"target_end_date\" columns have
             been excluded for readability. The \"output_type_id\" and \"value\"
             columns show the different predicted quantiles for the number of
             hospital admissions due to influenza made by the CMU-TimeSeries
             model for location 48 (Texas). The predictions were made during the
             week of December 16, 2023 (reference date) and are making a
             prediction for the horizon of 1 week ahead, for the week ending
             December 23.")
```
As the previous filter focus on only Texas, I thought maybe it might make it easier for the user to show horizon then location here. As horizon will be expected to change but not location.

LucieContamin commented on vignettes/hubverse_ms_vignette.Rmd:
```diff
@@ -0,0 +1,220 @@
+
+# Loading target data
+
+We can use the `aws.s3` package to read the contents of the most up-to-date target data file stored in the cloud.
```
Should it be updated to match the latest requirement/standard from hubverse? (to avoid rename, etc.)

LucieContamin replied:
```suggestion
We can use the `aws.s3` package to read the contents of the most up-to-date target data (also called "observed data") file stored in the cloud.
```
I suggest we add "observed data" because it's used in the next "visualization" part of the vignette.

LucieContamin commented on vignettes/hubverse_ms_vignette.Rmd:
```diff
@@ -0,0 +1,220 @@
+
+# Build ensemble and visualize
+
+Using the model output data collected above, we can pass these data to the `hubEnsembles::simple_ensemble()` function to build an ensemble that takes the mean of predicted values from each model at each quantile level. This new forecast output is then added to the existing forecasts and visualized all together. The hubEnsembles package supports a range of aggregation approaches including a linear ...
```
```suggestion
Using the model output data collected above, we can pass these data to the `hubEnsembles::simple_ensemble()` function to build an ensemble that takes the mean of predicted values from each model at each quantile level. This new forecast output is then added to the existing forecasts and visualized all together. The `hubEnsembles` package supports a range of aggregation approaches including a linear opinion pool (probability density averaging) or using the mean or median of quantiles.[@shandross_multi-model_2025] Model weights can be specified for the various approaches, allowing for some models to influence the output more strongly.
```
(following @eahowerton suggestion)

LucieContamin commented on vignettes/hubverse_ms_vignette.Rmd:
````diff
@@ -0,0 +1,220 @@
+
+Using the model output data collected above, we can pass these data to the `hubEnsembles::simple_ensemble()` function to build an ensemble that takes the mean of predicted values from each model at each quantile level. This new forecast output is then added to the existing forecasts and visualized all together. The hubEnsembles package supports a range of aggregation approaches including a linear ...
+
+```{r, fig.cap = "Forecasts of hospital admissions in Texas due to influenza in the 2023-2024 respiratory virus season. Predictions from five submitted models and an ensemble are shown along with observed data. The observed data are shown as a line with dots for specific weekly observations. Each individual model is shown in orange and the mean ensemble is shown in black, with the 80% prediction  ...
````
```suggestion
```{r, fig.cap = "Forecasts of hospital admissions in Texas due to influenza in the 2023-2024 respiratory virus season. Predictions from five submitted models and an ensemble are shown along with observed data. The observed data are shown as a line with dots for specific weekly observations. Each individual model is shown in gray and the mean ensemble is shown in orange, with the 80% prediction intervals shown as shaded regions."}
```
(following @eahowerton  suggestion)

LucieContamin commented on vignettes/hubverse_ms_vignette.Rmd:
```diff
@@ -0,0 +1,220 @@
+                                       group = "reference_date",
+                                       interactive = FALSE,
+                                       fill_transparency = .2,
+                                       intervals = 0.90,
```
```suggestion
                                       intervals = 0.80,
```

LucieContamin commented on vignettes/hubverse_ms_vignette.Rmd:
```diff
@@ -0,0 +1,220 @@
+
+# Evaluate forecasts
+
+The package hubEvals can be used to evaluate probabilistic predictions against observed data. The hubEvals package creates an interface for hubverse-style model output and target data to the scoringutils package, which is an engine for computing a number of common scores for probabilistic predictions.[@bosse-scoringutils-2024] 
```
```suggestion
The package `hubEvals` can be used to evaluate probabilistic predictions against observed data. The `hubEvals` package creates an interface for hubverse-style model output and target data to the `scoringutils` package, which is an engine for computing a number of common scores for probabilistic predictions.[@bosse-scoringutils-2024] 
```

LucieContamin commented on vignettes/hubverse_ms_vignette.Rmd:
````diff
@@ -0,0 +1,220 @@
+```
+
+# Software configuration
+The following demonstration shows how to access and analyze data from a real hub using `r R.version$string` and several of the hubverse R packages.[@the_consortium_of_infectious_disease_modeling_hubs_hubverse_2025] The following packages are on CRAN and may be installed using the standard `install.packages()` function: `hubEnsembles`, `aws.s3`, `arrow`, `ggplot2` and `dplyr`.
````
I also had to update some of the packages to have the vignette knitting, so it might be a good idea to indicate which minimal version are necessary for some packages to replicate the vignette (either at the beginning or the end of the document)

LucieContamin replied:
I understand, I am debating that question too. It's not necessary to install them because they are dependencies and install from hubverse packages, but we are still using them. So maybe I will not include the code to install them but make a note of the package used in the vignette.

nickreich replied:
I think this is addressed satisfactorily with @annakrystalli 's updates to the vignette.

LucieContamin commented on vignettes/hubverse_ms_vignette.Rmd:
```diff
@@ -0,0 +1,220 @@
+
+# Load forecast data
+CDC's FluSight Forecast Hub exists as a [GitHub repository](https://github.com/cdcepi/FluSight-forecast-hub).[@noauthor_cdcepiflusight-forecast-hub_2025] Using hubverse tools, it also has been mirrored to the cloud, so a copy of the data lives at a publicly-accessible AWS S3 bucket. This means that instead of needing to download a local copy of the repository, the data for this project can be dir ...
+
```
Should we add a little more context to the hub, saying what the goal is quickly and that the data mirrored on the bucket have all passed validation (using hubValidations)

LucieContamin reviewed (commented):
Thank you, it looks good and it all made sense to me. I was able to replicate (after some updates) the vignette. I made some minor comment and suggestions. Please feel free to let me know if any additional information is necessary.

annakrystalli reviewed (dismissed):
Thanks for pulling this out into a standalone hubverse vignette! 

I've added a number of comments, mainly for adhering to R convention a bit better but also a number of comments to add more context in a number of areas, mainly to help the vignette be able to indeed stand alone.

Will be making a separate PR to replace target data access code with `hubData` shortly!

annakrystalli committed:
Reorganise software requirement and installation section

annakrystalli committed:
Clean up Load forecast data section

annakrystalli committed:
Style and clean up plotting chunks

annakrystalli committed:
Update time series target data access section

annakrystalli committed:
Move silently loaded libraries to end, before ession info

annakrystalli committed:
Style code and remove unnecessary dplyr namespacing

annakrystalli committed:
Change the season to 2024-2025 to use oracle output data via hubData

annakrystalli committed:
remove unnecessary package loading

lshandross reviewed (changes requested):
The vignette is good in terms of giving code examples for how to access cloud data and analyze and plot it, but I agree with the others that it would benefit from a bit more explanation of why we might want to do the things demonstrated in each section.

nickreich committed:
Merge pull request #16 from hubverse-org/ak/hubverse-ms-vignette-edits

Ak/hubverse ms vignette edits

nickreich committed:
addressing some review comments.

nickreich committed:
updates to vignette, reflecting review comments.

lshandross reviewed (dismissed):
The updates all look great — thank you!

annakrystalli commented on vignettes/hubverse_ms_vignette.Rmd:
```diff
@@ -0,0 +1,331 @@
+
+# Load forecast data
+
+CDC's FluSight Forecast Hub exists as a [GitHub repository](https://github.com/cdcepi/FluSight-forecast-hub)[@noauthor_cdcepiflusight-forecast-hub_2025]. Using hubverse tools, the data for this hub has been validated and mirrored to the cloud, so a copy of the model output for this hub hub lives at a publicly-accessible AWS S3 bucket. This means that instead of needing to download a local copy of ...
```
```suggestion
CDC's FluSight Forecast Hub exists as a [GitHub repository](https://github.com/cdcepi/FluSight-forecast-hub)[@noauthor_cdcepiflusight-forecast-hub_2025]. Using hubverse tools, the data for this hub has been validated and mirrored to the cloud, so a copy of the model output for this hub lives at a publicly-accessible AWS S3 bucket. This means that instead of needing to download a local copy of the repository, the data for this project can be directly accessed from the cloud using `hubData` tools. The data is organised in a format we can open as an arrow dataset through R package `arrow` directly .
```

annakrystalli commented on vignettes/hubverse_ms_vignette.Rmd:
```diff
@@ -0,0 +1,331 @@
+The hub connection object printed above shows information about the hub, including the number and format of files loaded, the schema of the data, and column names and data types of model-output data. Note that the `skip_checks = TRUE` argument is used to skip the validation checks that are normally performed when connecting to a hub. In practice, since these data are all validated upon submission ...
+
+We then collect a subset of model outputs that we will interact with further in this demonstration. Specifically, we will extract zero through three week-ahead quantile predictions of weekly incident flu hospitalizations made for the state of Texas on three dates in late 2024 and early 2025. For simplicity, we will restrict our query to predictions made for five arbitrarily chosen models, althoug ...
+While the files are stored in the hub separately by model and reference date, the `hubData::collect_hub()` function, through functionality from `arrow`, allows us to act on all of the data collectively, as if if were part of a single `data.frame` object.
```
```suggestion
Although the data are stored in the hub across multiple files, organised by `model_id` and `round_id`, `hubData::connect_hub()` wraps functionality from `arrow` to open a connection to the data as an Arrow Dataset. This allows us to act on all of the data collectively, as if it were part of a single `data.frame` object.  Arrow datasets, including those stored on S3, behave as lazy data frames—they do not load data into memory until explicitly queried or collected. The `arrow` package also provides a [dplyr](https://dplyr.tidyverse.org/) interface to Arrow Datasets which allows us to query and manipulate the data.
```

annakrystalli commented on vignettes/hubverse_ms_vignette.Rmd:
````diff
@@ -0,0 +1,331 @@
+  ) |>
+  hubData::collect_hub()
+```
+
````
```suggestion
The `collect_hub` `hubData` function wraps the `dplyr::collect()` function but also coerces the output to a `model_out_tbl` class object where possible.
```

annakrystalli commented on vignettes/hubverse_ms_vignette.Rmd:
````diff
@@ -0,0 +1,331 @@
+ts_data_con <- hubData::connect_target_timeseries(hub_path)
+```
+
+Next, we filter the target data to the location and time period of interest. In this case, we are interested in the state of Texas (FIPS code 48) and the 2024-2025 respiratory virus season, which runs from September 23, 2024 to May 1, 2024. We also filter to the most recent `as_of` date to ensure we are using the latest available data. The `as_of` date is a metadata field that indicates when the  ...
````
```suggestion
Next, we filter the target data to the location and time period of interest. In this case, we are interested in the state of Texas (FIPS code 48) and the 2024-2025 respiratory virus season, which runs from September 23, 2024 to May 1, 2025. We also filter to the most recent `as_of` date to ensure we are using the latest available data. The `as_of` date is a metadata field that indicates when the target data was last updated.
```

annakrystalli commented on vignettes/hubverse_ms_vignette.Rmd:
````diff
@@ -0,0 +1,331 @@
+```{r}
+oo_data_con <- hubData::connect_target_oracle_output(hub_path)
+```
+
````
Note that the oracle-output contains a `horizon` column, which [as discussed](https://github.com/reichlab/decisions/pull/31#discussion_r2175444614), probably shouldn't be there?

I think we should resolve this or at least open an issue about resolving it soon in. I don't feel we should be showcasing wrongly formatted hubs in the vignette as it can cause unnecessary confusion.

annakrystalli committed:
Print target data connection objects

annakrystalli committed:
Remove unnecessary filtering. Use minimal observable unit.

This circumvents the fact the oracle output data contains an unnecessary `horizon` column, but should be revisited once the column is removed in the hub. https://github.com/hubverse-org/hubverse/pull/15/files#r2176783293

annakrystalli reviewed (commented):
I've made a few scattered suggestions but I've put my main outstanding issues in this PR https://github.com/hubverse-org/hubverse/pull/17

annakrystalli committed:
Update gitignore

annakrystalli committed:
Add additional context

nickreich committed:
Merge pull request #17 from hubverse-org/ak/hubverse-vignette/2

Updating oracle-output expectations

nickreich committed:
Merge branch 'main' into hubverse-ms-vignette

nickreich committed:
Apply suggestions from code review

Co-authored-by: Anna Krystalli <annakrystalli@googlemail.com>

nickreich committed:
add kable for oracle output and address lintr complaints

lshandross reviewed (dismissed):
Just fixed a couple of typos

nickreich committed:
Apply suggestions from code review

Co-authored-by: Li Shandross <57642277+lshandross@users.noreply.github.com>

nickreich committed:
adding udpated pdf.
